tests/versioned/koa.tap.js

```diff
@@ -0,0 +1,219 @@
+      )
+      var mid1 = trace.children[0]
+      t.equal(
+        mid1.name,
+        'Nodejs/Middleware/Koa/one',
+        '1st middleware segment has expected name'
+      )
+      var children = mid1.children
+      t.equal(children[0].name, 'Truncated/testSegment', 'trace has 1st test segment')
+      var mid2 = children[1]
+      t.equal(
+        mid2.name,
+        'Nodejs/Middleware/Koa/two',
+        '2nd middleware segment has expected name'
+      )
+      t.equal(children[2].name, 'Truncated/nestedSegment')
+      children = mid2.children
+      t.equal(children[0].name, 'timers.setTimeout', 'trace has 2nd test segment')
+      var mid3 = children[1]
+      t.equal(
+        mid3.name,
+        'Nodejs/Middleware/Koa/three',
+        '3rd middleware segment has expected name'
+      )
+      t.end()
+    })
+
+    run()
+  })
+
+  t.test('maintains transaction state between middleware', function(t) {
+    var tasks = []
+    var intervalId = setInterval(function() {
+      while (tasks.length) {
+        tasks.pop()()
+      }
+    }, 10)
+
+    t.tearDown(function() {
+      clearInterval(intervalId)
+    })
+    var tx
+
+    app.use(function one(ctx, next) {
+      tx = helper.agent.getTransaction()
+      return new Promise(executor)
+
+      function executor(resolve) {
+        tasks.push(function() {
+          next().then(function() {
+            t.transaction(tx)
+            resolve()
+          })
+        })
+      }
+    })
+    app.use(function two(ctx, next) {
+      t.transaction(tx)
+      return next()
+    })
+    app.use(function three(ctx) {
+      ctx.body = 'done'
+    })
+
+    helper.agent.on('transactionFinished', function(txn) {
+      checkMiddlewareSegments(t, txn)
+    })
+
+    run()
+  })
+
+  t.test('errors handled within middleware are not recorded', function(t) {
+    app.use(function one(ctx, next) {
+      return next().catch(function(err) {
+        t.equal(err.message, 'middleware error', 'caught expected error')
+      })
+    })
+    app.use(function two() {
+      throw new Error('middleware error')
+    })
+
+    helper.agent.on('transactionFinished', function(tx) {
+      var errors = helper.agent.errors.errors
+      t.equal(errors.length, 0, 'no errors are recorded')
+      checkMiddlewareSegments(t, tx)
+    })
+
+    run()
+  })
+
+  t.test('errors not handled in middleware are recorded', function(t) {
+    app.use(function one(ctx, next) {
+      return next()
+    })
+    app.use(function two() {
+      throw new Error('middleware error')
+    })
+    app.on('error', function(err) {
+      t.equal(err.message, 'middleware error', 'caught expected error')
+    })
+
+    helper.agent.on('transactionFinished', function(tx) {
+      var errors = helper.agent.errors.errors
+      t.equal(errors.length, 1, 'recorded expected number of errors')
+      var error = errors[0][2]
+      t.equal(error, 'middleware error', 'recorded expected error')
```

Comment:
This actually works!? I'm absolutely amazed this works given that the instrumentation doesn't specifically point it at anything. Going to have to investigate this.